### DISCUSSION TITLE
Can DataLoader's CacheMap achieve Thread isolation or request isolation?

### DISCUSSION THREAD

**Discussion by ChengHoHang** [graphql-java/graphql-java#3046] [Q&A]
i need to use `DataLoader` to proccess batch query, and i have some doubts as follow

## why i need to use cacheMap?

*  for my business, Query Duplicate Keys Each Request is a very common occurrence, cacheMap can reduce batch process keys.

## why i need to  achieve Thread isolation or request isolation
*  for my business, I need real-time queries, `cacheEnable` can reduce batch process keys from each request
*  but i see `CacheMap` is a thread-shared instance, so i can't new a `CacheMap` before query or clear Cache after query

In addition, i wanna know what's diffence bettween `CacheMap` and `ValueCache` in DataLoader

am i missing something? please give me some help, thanks @bbakerman

**Comment by ChengHoHang** [comment]
i refer to [doc](https://www.graphql-java.com/documentation/batching#per-request-data-loaders) it means per request should new a dataLoader...
![image](https://user-images.githubusercontent.com/48877754/206100704-419b433e-e611-4478-bab4-2557676dc80d.png)
`ValueCache` is a cache use for cross request cache value, and `CacheMap` is a cache only for dataLoader target, i have known what difference bettween them.


however, i think `DataLoader` and `DataLoaderRegistry` should be singleton like `DataFetcher`

**Comment by lllhhhyyy** [comment]
I have a similar situation too. For example, in order to make `org.dataloader.DataLoader#getStatistics` is meaningful, `Statistics` should be statistics cross request, `DataLoader` should be singleton :)

**Comment by bbakerman** [comment]
> however, i think DataLoader and DataLoaderRegistry should be singleton like DataFetcher

The fact is a DataLoader is a stateful object, even if you disregard the caches ie `CacheMap` and `ValueCache`

It has to be a stateful by design.  It tracks what completable futures have been asked for per "key" and then completes them when the batch function runs.  In other words it builds up state for each call to `load(k)` and then mutates that when the batch function runs and completes.

This is why it cant be a singleton - its very stateful.  And then as such you cant share them between requests because of that state

**Comment by bbakerman** [comment]
`DataLoaderRegistry` is an interesting one.  However since it gives out data loaders per "name" and those DataLoaders are per request stateful,  then by extension so is the `DataLoaderRegistry`
